azimuth/utils/ml/perturbation_functions.py

```diff
@@ -307,9 +323,17 @@ def typo(original: str, config: PerturbationTestingConfig) -> List[PerturbedUtte
         with contextlib.redirect_stdout(None):
             # While nlpaug fixes their useless print, we ignore it.
             perturbed_utterance = aug.augment(original, n=1)
-        # Issue 854 nac adds spaces between the hyphen and $ sign.
-        perturbed_utterance = perturbed_utterance.replace(" ' ", "'")
-        perturbed_utterance = perturbed_utterance.replace("$ ", "$")
+        # nac alters spacing around punctuation (e.g., around '; after $; before ?) and replaces
+        # apostrophes with quotes; revert to original.
+        re_punctuation = re.compile(r"(\w+)")
+        punctuations = re_punctuation.split(original)[::2]
+        words = re_punctuation.split(perturbed_utterance)[1::2]
+        perturbed_utterance = "".join(
+            [
+                punctuation + word
+                for punctuation, word in zip_longest(punctuations, words, fillvalue="")
+            ]
+        )
```

Comment:
We don't need a list comprehension here, `str.join()` supports the generator directly.
```suggestion
        perturbed_utterance = "".join(
            punctuation + word
            for punctuation, word in zip_longest(punctuations, words, fillvalue="")
        )
```